### ISSUE TITLE
[2026/2/4]推薦・機械学習勉強会

### ISSUE THREAD

**Issue by Hayashi-Yudai** [wantedly/machine-learning-round-table#332]
## Why

推薦・機械学習勉強会は、推薦や機械学習、その周辺技術を通じてサービスを改善することにモチベーションのある人達の集まりです。ニュースやブログから論文まで、気になったものについてお互い共有しましょう！

発信のため、ここは **public** にしてあります。外部からの参加をご希望の方は[林/python_walker](https://x.com/python_walker)、[市村/chimuichimu](https://x.com/chimuichimu1) まで DM を送るか、Wantedly Visit の募集（https://www.wantedly.com/projects/391912） よりご連絡ください！

## What

Wantedly では隔週水曜日に

- 推薦の評価指標について議論したい
- 〇〇っていうライブラリ / フレームワークを導入してみたい
- 他社の基盤事例をみんなにシェアして自社の基盤開発に活かしたい
- もっと推薦を良くするためにどんなものが必要か議論したい

といった話をする「推薦・機械学習勉強会」を開催しています。
この ISSUE はその会で話すネタを共有するための場所です。

話したいことがある人はここにコメントしましょう！
会の間に話した内容もここにメモしましょう！

prev: https://github.com/wantedly/machine-learning-round-table/issues/330

**Comment by taikinman** [comment]
## [Towards Generalizable and Efficient Large-Scale Generative Recommenders](https://netflixtechblog.medium.com/towards-generalizable-and-efficient-large-scale-generative-recommenders-a7db648aa257)

### 概要

生成型推薦モデルの大規模化において発生する以下の課題に対するNetflixのアプローチを紹介したブログ。

1. モデルサイズとパフォーマンスの関係性の理解
2. 学習・推論効率
3. サービング遅延と順序不変性
4. 未見アイテム・コールドスタート問題

#### 0. 検証設定

タスクA：ユーザーの長期嗜好が支配的な行動を当てる推薦タスク。次にAを見るかBを見るかの順序が入れ替わり得るので、次の行動をピンポイントで当てるのが難しい。
タスクB：直近の行動の直後に何を出すかが重要な短期依存が強い推薦タスク。長期嗜好に比べて候補や順序の幅が狭いが、行動ログ依存なので難易度中。
タスクC：時間依存の行動に関する推薦タスク。

#### 1. モデルサイズとパフォーマンスの関係性の理解

LLMにおけるスケーリング則としてはChinchilla則やlog-scale則などがあるが、推薦タスクではタスク依存かつ性能上限を設定した方がより良いフィッティングが得られた

$$
P_0 - P = (N / N_0)^{-a}
$$

$P_0$ : 特定のタスクで達成可能な性能の上限値（タスク自体の難易度で変わる）
→ $P_0$ はタスクの難易度を表す指標となり得る

<img width="595" height="355" alt="image" src="https://github.com/user-attachments/assets/09b221f4-3475-4da6-a8fb-42fcc8c543f6" />


指標をMRR（Mean Reciprocal Rank）とすると、タスクAでは、 $P_0=0.311$ で性能が頭打ちになりそうな一方、タスクCでは $P_0=1.075$ でモデルサイズを大きくすれば完璧に正解できると思われる。

→大規模化による運用コスト増と精度向上/体験向上のトレードオフが予測できる？

#### 2. 学習・推論効率

生成型推薦タスクでは、アイテムID or (アイテム x イベント)を語彙として扱い、次のトークンの予測を行う。LLMの数十倍の語彙が必要（GPT-3の語彙の40倍）。デコーダーで素朴に全語彙のsoftmaxを取ると計算量爆発。また、定期的に2兆トークンを処理する（再学習など？）。色々工夫しているが、特にデコーディングの効率化で大きな効果があった。

工夫：
1. sampled softmax：負例をサンプリングしてそのサブセットで分類問題を解く 
2. projected head：出力層でlogits計算前に埋め込み次元を低次元に射影して計算量削減

→語彙サイズが大きい領域で学習コストを1〜2桁削減

<img width="393" height="260" alt="image" src="https://github.com/user-attachments/assets/8f456eb0-6faa-4be2-9ff4-46103ce7e061" />

Netflixのモデルは80台のA100 GPU（各80GBメモリ）で周期的に240時間のトレーニングを実施して頻繁にファインチューニングしているらしい。

#### 3. サービング遅延と順序不変性

推薦の出力はしばしばキャッシュされ、ユーザーの新しい行動から数分～数日遅れて配信されるため、学習時に次に見るはずのアイテムAを当てても、実際の配信辞典ではユーザーがすでにアイテムAを消費していてアイテムBを配信するのが妥当というズレが起きがち。
また、推薦では順序不変性（あるユーザーが次にAとBの両方を視聴する可能性が高い場合、順序（A→BまたはB→A）はどちらでも良い）があるため、単純なNext Token Predictionでは正解候補を不当に罰してしまう（オンライン指標とオフライン指標のギャップにもなり得る）。

工夫：
1. 次のトークンを1つ当てるのではなく、可能性のある複数の将来アイテム集合を教師ラベルとし、この集合に入っているアイテム群を高確率にするように学習するMulti Token Predictionを採用
    - 損失はコンテキストに時間的に近いものを重視し、アイテムの価値（視聴時間、完了率、新規性、多様性）でも重みづけ

**MTPの効果**

オンラインサービス環境（95パーセンタイル遅延<1秒）と低コスト・長時間遅延対応サービス（遅延時間48時間）でタスクA, B, Cの評価

- MTPでサービング遅延が長い場合の精度が向上
- 短期的な依存関係が強いタスクではNTPよりわずかに悪化

<img width="595" height="245" alt="image" src="https://github.com/user-attachments/assets/18f8e868-6f89-4921-98ab-4a932eeb139d" />

#### 4. 未見アイテム・コールドスタート問題

学習と推論に時間を要する大規模推薦モデルは、事前学習データに含まれていない新規コンテンツに対しても適切に理解し推薦出来る能力が要求される。

工夫：
1. 視覚情報、言語情報、知識グラフを統合したマルチモーダルアイテムタワー（Semantic Item Tower）を採用
    - それぞれの情報を専用モデルでエンコードし、生成型推薦モデルの学習可能なプロジェクタに注入
2. 学習可能な埋め込み表現（協調フィルタリング情報）とSemantic Item Towerによる埋め込み表現を重みづけ和
    - 学習時には協調フィルタリング埋め込みをランダムにマスクしてOOVトークンの埋め込みで置き換え
    - Semantic Item Tower由来の埋め込みをembedding layerだけでなくdense layer weightsにも注入

→意味的情報を利用するようになり、未見アイテムやコールドスタートアイテムへの対応能力が向上

<img width="466" height="225" alt="image" src="https://github.com/user-attachments/assets/bb833b77-06b2-43be-9db4-215a54f520c9" />

### モデルの大規模化による効果

モデルサイズを50Mから1Bにスケールアップしてパフォーマンスを評価

- 全てのタスクで1Bモデルが50Mよりいいパフォーマンス

<img width="308" height="238" alt="image" src="https://github.com/user-attachments/assets/2cfa3fbd-76d6-468b-a97d-3c88e9c5dc55" />

**Comment by wakama1994** [comment]
### [ゼロから始める勾配ブースティング決定木の理論](https://zenn.dev/dalab/articles/9c843f0ec8aabf)
- LightGBMを業務でしっかり目に扱いそうなため、事前学習
- 補足として下記も参照しました
  - [LightGBM 徹底入門 – LightGBMの使い方や仕組み、XGBoostとの違いについて](https://www.codexa.net/lightgbm-beginner/)
   - [LightGBMとは：高効率な勾配ブースティングフレームワークの概要](https://www.issoh.co.jp/tech/details/3919/)
- 追記 質疑の際にコメントであった記事
  - https://www.slideshare.net/slideshow/miru2020-tutorial-237272385/237272385
  - https://speakerdeck.com/hakubishin3/target-encoding-hanazeyou-xiao-nafalseka

**Comment by Hayashi-Yudai** [comment]
## [Beyond Two Towers: Re-architecting the Serving Stack for Next-Gen Ads Lightweight Ranking Models (Part 1)](https://medium.com/pinterest-engineering/beyond-two-towers-re-architecting-the-serving-stack-for-next-gen-ads-lightweight-ranking-models-1992f2b76cbb)

Pinterest での候補集合の計算時のアーキテクチャを刷新するプロジェクトの話。モデルの話というよりもインフラ周りの話が多め。

### 課題感

Two tower に限界を感じている。高速に候補を計算できる点は魅力的だが、インタラクション情報を扱うのが構造上難しいのがネックだった。

### 解決方法

候補集合の計算用に GPU の推論環境を整備した。単純に GPU に置き換えれば全て解決する、というわけでも無く、様々な工夫を組み合わせることで応答性を高めた。

#### 特徴量取得の高速化

- 頻繁に出てくるアイテムの特徴量は GPU メモリ上に載せておく (モデルパラメーターと同じような扱い)
- 他のアイテムに関しては高速な KV-Cache を利用

#### ビジネスロジック計算の高速化

- ビジネスロジックの計算をモデルの中で行うようにした (テンソルのまま GPU 上で計算したということらしい)

ここまでの最適化でも、応答時間の p90 は 4,000 sec だったらしく、更なる高速化をした。

#### 推論の高速化

- 推論時に CUDA 上で Multi-stream を利用して計算
- Worker スレッド数を CPU のスレッド数と同じに揃えた。かつ、処理をするコアを固定した (OSによるコンテキストスイッチの防止)
- Triton Kernel を利用して、頻出するレイヤー同士の組み合わせを fusion。メモリ書き込み数を減らして高速化
- BF16 の利用

**Comment by nogawanogawa** [comment]
## [マルチモーダルLLMを活用したZOZOTOWN検索の関連性評価手法](https://techblog.zozo.com/entry/search-quantitative-evaluation-llm)

- LLMによる判定：大規模言語モデルが関連性を自動判定
  - 大量データの処理が可能
  - 一貫性のある判定基準
  - コスト効率の良さ
  - 柔軟な基準適用
  - バイアスの軽減
- プロンプト設計が重要
  - Role
  - Description
  - Narrative
  - Aspects
  - Output

<img width="720" height="235" alt="Image" src="https://github.com/user-attachments/assets/901e32c4-b44b-4b5b-b89e-7054e85a5c45" />

**Comment by chimuichimu** [comment]
## [リクルート『ホットペッパーグルメ』が Amazon OpenSearch Service で Hybrid Search を実現し検索体験を革新](https://aws.amazon.com/jp/blogs/news/amazon-opensearch-service-recruit-hpg/)

**Comment by ghibney** [comment]
## [ヤフートップページの裏側：記事推薦システムの試行錯誤と今後の挑戦](https://techblog.lycorp.co.jp/ja/20260202a)
* Yahoo! JAPANトップのニュース推薦に関する記事（記事内容自体は少し古い）
* BERTやRNNを用いた記事・ユーザーのベクトル学習により精度向上を図りつつ、多様性確保やdislike対応に課題があった
* 分布を考慮したリランキングや乱数初期値の影響対策など、実運用での試行錯誤と今後の改善方針を紹介
